Add Gitlab support

rarkins opened the pull request:
This PR adds GitLab support to close #65.

To configure it:

1. Set repository `platform` configuration field - this defaults to "github" so must be explicitly set to "gitlab" either at the global config level or per-repository level.
2. Configure a `token` via standard `renovate` config options, or have `GITLAB_TOKEN` as an environment variable
3. You probably still need a GitHub token so as not to exhaust the anonymous API limit due to change request retrievals

Still to do:

- [x] Testing/validation
- [x] Docs
- [x] Use stable `gl-got`

rarkins commented:
I just realised that GitLab users will probably still want a GITHUB_TOKEN defined, to avoid rate limiting when checking GitHub changelogs via their API

hbetts commented on lib/api/gitlab.js:
```diff
@@ -0,0 +1,296 @@
+  if (token) {
+    process.env.GITLAB_TOKEN = token;
+  } else if (!process.env.GITLAB_TOKEN) {
+    throw new Error(`No token found for GitHub repository ${repoName}`);
```
`GitHub` => `GitLab`

rarkins replied:
Fixing

hbetts commented on lib/api/gitlab.js:
```diff
@@ -0,0 +1,296 @@
+};
+
+// Initialize GitLab by getting base branch
+async function initRepo(repoName, token) {
```
Line _45_ of `index.js` does not pass the `token` from `config.token` to this function:

```javascript
await api.initRepo(config.repository);
```

rarkins replied:
Fixing

hbetts commented on lib/api/gitlab.js:
```diff
@@ -0,0 +1,296 @@
+async function branchExists(branchName) {
+  logger.debug(`Checking if branch exists: ${branchName}`);
+  try {
+    const url = `projects/${config.repoName}/repository/branches/${branchName}`;
```
All uses of `gl-got` expect that the project is on `gitlab.com`.

It looks like I can set `GITLAB_ENDPOINT` environment variable.

Could that environment variable be exposed through a CLI argument, perhaps `--endpoint`, similar to how `token` is exposed as `--token` and then used to set the `TOKEN` environment variable?

rarkins replied:
Good idea - done

hbetts commented on lib/api/gitlab.js:
```diff
@@ -0,0 +1,296 @@
+
+async function addLabels(prNo, labels) {
+  logger.debug(`Adding labels ${labels} to #${prNo}`);
+  let url = `projects/${config.repoName}/merge_requests/prNo`;
```
I realized what the error was I was receiving. The `prNo` string needs to be changed to `${prNo}`.

rarkins replied:
Done. same with assignees

hbetts commented:
The `configureRepository` function that generates the _Configure Renovate_ merge request is creating a configuration file with the `platform` value set to `github`, instead of `gitlab` as specified in the `platform` cli option passed to `renovate`. `renovate` is calling `defaultParser.getConfig()`, which is pulling the configuration values, including `platform`, from here - https://github.com/singapore/renovate/blob/master/lib/config/definitions.js#L22

rarkins commented:
@hbetts Now fixed `configureRepository` for gitlab use

singapore commented:
This branch needs a new `commitFilesToBranch` API added, which will be utilised once the yarn branch is merged.

rarkins commented:
@hbetts you should be able to test `yarn` support in this branch now

hbetts commented:
Thank you @rarkins!

I ran `renovate` over my test project again.

I ran into an issue when `renovate` called `getChangelog`, which calls `changelog.generate`.

```bash
debug: Creating branch renovate/semantic-release-gitlab-3.x
debug: package.json exists - updating it
debug: yarn.lock exists - updating it
debug: semverString: >2.5.2 <=3.0.0
verbose: getChangelog error: Unknown Github Repo: git+https://gitlab.com/hyper-expanse/semantic-release-gitlab.git. Not Found
verbose: No changelog for semantic-release-gitlab
```

It does not seem that `changelog` supports anything but GitHub - https://github.com/dylang/changelog

There was also a minor issue with the markdown. When it attempted to render `^2.5.1`, it rendered it in superscript form.

![markdown](https://cloud.githubusercontent.com/assets/6701030/22851564/badbd8cc-efeb-11e6-8fe3-bd37ede419af.png)

Looks like [GitLab uses](https://docs.gitlab.com/ee/user/markdown.html) the `redcarpet` gem package for markdown rendering. Their [documentation for superscript] seems to indicate that `^` is notation for making text after `^` superscript.

Except for those minor issues `renovate` is working great.

hbetts commented:
The only closest thing I've found is [changed-log](https://www.npmjs.com/package/changed-log). However, it also does [not support GitLab](https://github.com/bahmutov/changed-log/issues/31).

rarkins commented:
I've added https://github.com/dylang/changelog/issues/29 for (one day) supporting GitLab repos in Changelog.

Also wrapped versions in backticks for the `prBody` field which should work around the GitLab superscript problem in the first line at least.

Finally, I've added #97 to consider switching from `changelog` to `changed-log` if it's better.

rarkins commented:
Published to npm as version 8.7.0
Thanks to @hbetts for advice and testing
